Overloaded setters not allowed on planning entity in Solver 2.0

Consider a planning entity `SolverStop` with the following property and two setters:
```
    private MultipleTimeWindow multipleTimeWindow;

    public void setMultipleTimeWindow(MultipleTimeWindow multipleTimeWindow) {
        this.multipleTimeWindow =
                multipleTimeWindow == null ? new MultipleTimeWindow() : multipleTimeWindow;
    }

    public void setMultipleTimeWindow(List<TimeWindow> timeWindows) {
        this.multipleTimeWindow = new MultipleTimeWindow(timeWindows);
    }
```

Solver fails to generate a solution cloner with the following exception:

```
[ERROR] Caused by: java.lang.IllegalStateException: The containingClass (class ai.timefold.model.pickupdeliveryrouting.solver.domain.SolverStop) has multiple setter methods ([public void ai.timefold.model.pickupdeliveryrouting.solver.domain.SolverStop.setMultipleTimeWindow(ai.timefold.model.pickupdeliveryrouting.solver.domain.MultipleTimeWindow), public void ai.timefold.model.pickupdeliveryrouting.solver.domain.SolverStop.setMultipleTimeWindow(java.util.List)]) with the propertyName (multipleTimeWindow).                                                                                                           
[ERROR]         at ai.timefold.solver.core.impl.domain.common.ReflectionHelper.getSetterMethod(ReflectionHelper.java:263)
[ERROR]         at ai.timefold.solver.core.impl.domain.common.accessor.gizmo.GizmoFieldHandler.<init>(GizmoFieldHandler.java:30)
[ERROR]         at ai.timefold.solver.core.impl.domain.common.accessor.gizmo.GizmoMemberHandler.of(GizmoMemberHandler.java:32)
[ERROR]         at ai.timefold.solver.core.impl.domain.common.accessor.gizmo.GizmoMemberDescriptor.<init>(GizmoMemberDescriptor.java:126)
[ERROR]         at ai.timefold.solver.quarkus.deployment.GizmoMemberAccessorEntityEnhancer.createMemberDescriptorForField(GizmoMemberAccessorEntityEnhancer.java:401)
[ERROR]         at ai.timefold.solver.quarkus.deployment.GizmoMemberAccessorEntityEnhancer.getGizmoSolutionOrEntityDescriptorForEntity(GizmoMemberAccessorEntityEnhancer.java:366)
[ERROR]         at ai.timefold.solver.quarkus.deployment.GizmoMemberAccessorEntityEnhancer.lambda$generateSolutionCloner$0(GizmoMemberAccessorEntityEnhancer.java:282)
[ERROR]         at io.quarkus.gizmo2.impl.GizmoImpl.lambda$class_$1(GizmoImpl.java:134)
[ERROR]         at io.smallrye.classfile.impl.ClassFileImpl.build(ClassFileImpl.java:144)
[ERROR]         at io.smallrye.classfile.ClassFile.build(ClassFile.java:558)
[ERROR]         at io.quarkus.gizmo2.impl.GizmoImpl.class_(GizmoImpl.java:131)
[ERROR]         at io.quarkus.gizmo2.Gizmo.class_(Gizmo.java:89)
[ERROR]         at ai.timefold.solver.quarkus.deployment.GizmoMemberAccessorEntityEnhancer.generateSolutionCloner(GizmoMemberAccessorEntityEnhancer.java:251)
[ERROR]         at ai.timefold.solver.quarkus.deployment.TimefoldProcessor.lambda$generateDomainAccessors$6(TimefoldProcessor.java:1092)
```

--- a/core/src/main/java/ai/timefold/solver/core/config/util/ConfigUtils.java
+++ b/core/src/main/java/ai/timefold/solver/core/config/util/ConfigUtils.java
@@ -618,9 +618,9 @@ public class ConfigUtils {
             return null;
         }
         if (methods.length > 1) {
-            throw new IllegalStateException("The containingClass (" + containingClass
-                    + ") has multiple setter methods (" + Arrays.toString(methods)
-                    + ") with the propertyName (" + propertyName + ").");
+            throw new IllegalStateException(
+                    "The containingClass (%s) has multiple setter methods (%s) with the propertyName (%s)."
+                            .formatted(containingClass, Arrays.toString(methods), propertyName));
         }
         return methods[0];
     }

--- a/core/src/main/java/ai/timefold/solver/core/config/util/ConfigUtils.java
+++ b/core/src/main/java/ai/timefold/solver/core/config/util/ConfigUtils.java
@@ -101,7 +101,7 @@ public class ConfigUtils {
         }
         var beanClass = bean.getClass();
         customProperties.forEach((propertyName, valueString) -> {
-            var setterMethod = ReflectionHelper.getSetterMethod(beanClass, propertyName);
+            var setterMethod = getSetterMethod(beanClass, propertyName);
             if (setterMethod == null) {
                 throw new IllegalStateException(
                         """
@@ -604,11 +604,31 @@ public class ConfigUtils {
         return value.toString();
     }
 
+    /**
+     * @param containingClass never null
+     * @param propertyName never null
+     * @return null if it doesn't exist
+     */
+    private static Method getSetterMethod(Class<?> containingClass, String propertyName) {
+        String setterName = ReflectionHelper.PROPERTY_MUTATOR_PREFIX + ReflectionHelper.capitalizePropertyName(propertyName);
+        Method[] methods = Arrays.stream(containingClass.getMethods())
+                .filter(method -> method.getName().equals(setterName))
+                .toArray(Method[]::new);
+        if (methods.length == 0) {
+            return null;
+        }
+        if (methods.length > 1) {
+            throw new IllegalStateException("The containingClass (" + containingClass
+                    + ") has multiple setter methods (" + Arrays.toString(methods)
+                    + ") with the propertyName (" + propertyName + ").");
+        }
+        return methods[0];
+    }
+
     // ************************************************************************
     // Private constructor
     // ************************************************************************
 
     private ConfigUtils() {
     }
-
 }

--- a/core/src/main/java/ai/timefold/solver/core/impl/domain/common/ReflectionHelper.java
+++ b/core/src/main/java/ai/timefold/solver/core/impl/domain/common/ReflectionHelper.java
@@ -36,7 +36,7 @@ public final class ReflectionHelper {
             PROPERTY_ACCESSOR_PREFIX_IS
     };
 
-    private static final String PROPERTY_MUTATOR_PREFIX = "set";
+    public static final String PROPERTY_MUTATOR_PREFIX = "set";
 
     /**
      * Returns the JavaBeans property name of the given member.
@@ -172,7 +172,7 @@ public final class ReflectionHelper {
         return null;
     }
 
-    private static String capitalizePropertyName(String propertyName) {
+    public static String capitalizePropertyName(String propertyName) {
         if (propertyName.isEmpty() || Character.isUpperCase(propertyName.charAt(0))) {
             return propertyName;
         } else {
@@ -243,27 +243,6 @@ public final class ReflectionHelper {
     public static Method getDeclaredSetterMethod(Class<?> containingClass, Class<?> propertyType, String propertyName) {
         String setterName = PROPERTY_MUTATOR_PREFIX + capitalizePropertyName(propertyName);
         return getDeclaredMethod(containingClass, setterName, propertyType);
-    }
-
-    /**
-     * @param containingClass never null
-     * @param propertyName never null
-     * @return null if it doesn't exist
-     */
-    public static Method getSetterMethod(Class<?> containingClass, String propertyName) {
-        String setterName = PROPERTY_MUTATOR_PREFIX + capitalizePropertyName(propertyName);
-        Method[] methods = Arrays.stream(containingClass.getMethods())
-                .filter(method -> method.getName().equals(setterName))
-                .toArray(Method[]::new);
-        if (methods.length == 0) {
-            return null;
-        }
-        if (methods.length > 1) {
-            throw new IllegalStateException("The containingClass (" + containingClass
-                    + ") has multiple setter methods (" + Arrays.toString(methods)
-                    + ") with the propertyName (" + propertyName + ").");
-        }
-        return methods[0];
     }
 
     public static boolean isMethodOverwritten(Method parentMethod, Class<?> childClass) {

--- a/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/MemberAccessorFactory.java
+++ b/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/MemberAccessorFactory.java
@@ -81,7 +81,7 @@ public final class MemberAccessorFactory {
         if (member instanceof Field field) {
             var getter = ReflectionHelper.getGetterMethod(field.getDeclaringClass(), field.getName());
             if (getter == null) {
-                var setter = ReflectionHelper.getSetterMethod(field.getDeclaringClass(), field.getName());
+                var setter = ReflectionHelper.getSetterMethod(field.getDeclaringClass(), field.getType(), field.getName());
                 if (setter != null) {
                     throw new IllegalArgumentException("%s field (%s) on class (%s) has a setter (%s) but no getter."
                             .formatted(messagePrefix, field.getName(), field.getDeclaringClass().getCanonicalName(), setter));

--- a/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/MemberAccessorValidator.java
+++ b/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/MemberAccessorValidator.java
@@ -156,8 +156,9 @@ final class MemberAccessorValidator {
     private static void verifyIsPublicFieldOrHasPublicSetter(Member member, String messagePrefix) {
         switch (member) {
             case Field field -> {
+                var memberType = field.getType();
                 var getterMethod = ReflectionHelper.getGetterMethod(field.getDeclaringClass(), field.getName());
-                var setterMethod = ReflectionHelper.getSetterMethod(field.getDeclaringClass(), field.getName());
+                var setterMethod = ReflectionHelper.getSetterMethod(field.getDeclaringClass(), memberType, field.getName());
                 if (setterMethod == null) {
                     if (!Modifier.isPublic(field.getModifiers())) {
                         throw new IllegalArgumentException(
@@ -174,10 +175,11 @@ final class MemberAccessorValidator {
                 }
             }
             case Method getterMethod -> {
+                var memberType = getterMethod.getReturnType();
                 verifyGetterName(getterMethod, messagePrefix);
                 var memberName = ReflectionHelper.getGetterPropertyName(getterMethod);
                 var setterMethod = ReflectionHelper.getSetterMethod(getterMethod.getDeclaringClass(),
-                        memberName);
+                        memberType, memberName);
                 if (setterMethod == null) {
                     throw new IllegalArgumentException("""
                             %s requires both a public getter and a public setter but only have a public getter.

--- a/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/gizmo/GizmoFieldHandler.java
+++ b/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/gizmo/GizmoFieldHandler.java
@@ -26,8 +26,9 @@ final class GizmoFieldHandler implements GizmoMemberHandler {
     GizmoFieldHandler(Class<?> declaringClass, FieldDesc fieldDescriptor, boolean canBeWritten) {
         this.declaringClass = declaringClass;
         this.fieldDescriptor = fieldDescriptor;
+        var fieldType = GizmoMemberHandler.resolveType(declaringClass, fieldDescriptor.type());
         var getterMethod = ReflectionHelper.getGetterMethod(declaringClass, fieldDescriptor.name());
-        var setterMethod = ReflectionHelper.getSetterMethod(declaringClass, fieldDescriptor.name());
+        var setterMethod = ReflectionHelper.getSetterMethod(declaringClass, fieldType, fieldDescriptor.name());
 
         if (getterMethod == null) {
             getterDescriptor = null;

--- a/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/gizmo/GizmoMemberDescriptor.java
+++ b/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/gizmo/GizmoMemberDescriptor.java
@@ -232,8 +232,9 @@ public final class GizmoMemberDescriptor {
     }
 
     private static Optional<MethodDesc> lookupSetter(Object memberDescriptor, Class<?> declaringClass, String name) {
-        if (memberDescriptor instanceof MethodDesc) {
-            return Optional.ofNullable(ReflectionHelper.getSetterMethod(declaringClass, name))
+        if (memberDescriptor instanceof MethodDesc methodDesc) {
+            var returnType = GizmoMemberHandler.resolveType(declaringClass, methodDesc.returnType());
+            return Optional.ofNullable(ReflectionHelper.getSetterMethod(declaringClass, returnType, name))
                     .map(MethodDesc::of);
         } else {
             return Optional.empty();

--- a/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/gizmo/GizmoMemberHandler.java
+++ b/core/src/main/java/ai/timefold/solver/core/impl/domain/common/accessor/gizmo/GizmoMemberHandler.java
@@ -1,6 +1,8 @@
 package ai.timefold.solver.core.impl.domain.common.accessor.gizmo;
 
 import java.lang.constant.ClassDesc;
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Array;
 import java.lang.reflect.Field;
 import java.lang.reflect.Method;
 import java.lang.reflect.Modifier;
@@ -81,5 +83,36 @@ interface GizmoMemberHandler {
         // "L" + className + ";"
         var classDescriptor = classDesc.descriptorString();
         return classDescriptor.substring(1, classDescriptor.length() - 1).replace('/', '.');
+    }
+
+    static Class<?> resolveType(Class<?> declaringClass, ClassDesc classDesc) {
+        try {
+            return classDesc.resolveConstantDesc(MethodHandles.publicLookup().in(declaringClass));
+        } catch (ReflectiveOperationException e) {
+            // Probably Quarkus, due to different class loaders
+            if (!classDesc.isArray()) {
+                try {
+                    return Thread.currentThread().getContextClassLoader().loadClass(getTypeName(classDesc));
+                } catch (ClassNotFoundException ex) {
+                    throw new IllegalStateException("Unable to resolve type (%s) referenced from class (%s).".formatted(
+                            getTypeName(classDesc), declaringClass.getCanonicalName()), ex);
+                }
+            } else {
+                var componentType = resolveType(declaringClass, classDesc.componentType());
+                var arrayDepth = 0;
+                var currentIndex = 0;
+
+                // No convenience method to get array dimensions
+                var descriptorText = classDesc.descriptorString();
+                while (currentIndex != -1) {
+                    currentIndex = descriptorText.indexOf('[', currentIndex);
+                    arrayDepth++;
+                    if (currentIndex != -1) {
+                        currentIndex++;
+                    }
+                }
+                return Array.newInstance(componentType, arrayDepth).getClass();
+            }
+        }
     }
 }

--- a/core/src/test/java/ai/timefold/solver/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerTest.java
+++ b/core/src/test/java/ai/timefold/solver/core/impl/domain/solution/cloner/gizmo/GizmoSolutionClonerTest.java
@@ -104,7 +104,7 @@ class GizmoSolutionClonerTest extends AbstractSolutionClonerTest {
                                 AccessorInfo.withReturnValueAndNoArguments());
                     } else {
                         var getter = ReflectionHelper.getGetterMethod(currentClass, field.getName());
-                        var setter = ReflectionHelper.getSetterMethod(currentClass, field.getName());
+                        var setter = ReflectionHelper.getSetterMethod(currentClass, field.getType(), field.getName());
                         if (getter != null && setter != null) {
                             var getterDescriptor = MethodDesc.of(field.getDeclaringClass(),
                                     getter.getName(),

--- a/core/src/test/java/ai/timefold/solver/core/testdomain/reflect/field/TestdataFieldAnnotatedEntity.java
+++ b/core/src/test/java/ai/timefold/solver/core/testdomain/reflect/field/TestdataFieldAnnotatedEntity.java
@@ -42,6 +42,10 @@ public class TestdataFieldAnnotatedEntity extends TestdataObject {
         this.value = value;
     }
 
+    public void setValue(String code) {
+        this.value = new TestdataValue(code);
+    }
+
     // ************************************************************************
     // Complex methods
     // ************************************************************************

--- a/core/src/test/java/ai/timefold/solver/core/testdomain/reflect/generic/TestdataGenericEntity.java
+++ b/core/src/test/java/ai/timefold/solver/core/testdomain/reflect/generic/TestdataGenericEntity.java
@@ -50,6 +50,10 @@ public class TestdataGenericEntity<T> extends TestdataObject {
         this.value = value;
     }
 
+    public void setValue(String code) {
+        this.value = new TestdataGenericValue<>(code);
+    }
+
     public void setSubTypeValue(TestdataGenericValue<T> subTypeValue) {
         this.subTypeValue = subTypeValue;
     }

